New `Encamina.Enmarcha.AspNet.OpenApi` package and Improved memory store event types.

rliberoff opened the pull request:
- New `preview-07` for version `8.1.3`.
- Added new project `Encamina.Enmarcha.AspNet.OpenApi` with goodies for OpenAPI.
- Improved memory store event types, when they are raised and the data inside the arguments.
- Fixed some types and `README.md` issues.

rliberoff committed:
- Improved memory store event types, when they are raised and the data inside the arguments.

rliberoff committed:
- New `preview-07` for version `8.1.3`.
- Added new project `Encamina.Enmarcha.AspNet.OpenApi` with goodies for OpenAPI.
-
- Fixed some types and `README.md` issues.

rliberoff committed:
Minor typo fix.

fjbelizon commented on src/Encamina.Enmarcha.AspNet.OpenApi.Swashbuckle/README.md:
```diff
@@ -2,7 +2,7 @@
 [![Nuget package](https://img.shields.io/nuget/v/Encamina.Enmarcha.AspNet.OpenApi.Swashbuckle)](https://www.nuget.org/packages/Encamina.Enmarcha.AspNet.OpenApi.Swashbuckle)
 
-The Data Qdrant project contains functionalities related to the [OpenApi Swashbuckle](https://learn.microsoft.com/es-es/aspnet/core/tutorials/web-api-help-pages-using-swagger?view=aspnetcore-7.0)
+This package provides extensions and filters for OpenAPI (Swagger) with version support using [Swashbuckle](https://learn.microsoft.com/es-es/aspnet/core/tutorials/web-api-help-pages-using-swagger?view=aspnetcore-7.0).
```
The link should be to .NET 8 --> https://learn.microsoft.com/es-es/aspnet/core/tutorials/web-api-help-pages-using-swagger?view=aspnetcore-8.0

rliberoff replied:
Fixed.

fjbelizon commented on src/Encamina.Enmarcha.AspNet.OpenApi/Extensions/IApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <param name="app">A valid instance of <see cref="IApplicationBuilder"/> as the application's request pipeline builder.</param>
+    /// <param name="setupAction">An optional setup action for the <see cref="GroupNameKeyAuthenticationOptions"/>.</param>
+    /// <returns>The <see cref="IApplicationBuilder"/> so that additional calls can be chained.</returns>
+    public static IApplicationBuilder UseOpenApiGroupNameKeyAuthorization(this IApplicationBuilder app, Action<GroupNameKeyAuthenticationOptions> setupAction = null)
```
This method should be overloaded, do not use optional parameters.

rliberoff replied:
Fixed.

fjbelizon commented on src/Encamina.Enmarcha.SemanticKernel/MemoryStoreExtender.cs:
```diff
@@ -61,7 +61,7 @@ public virtual async Task DeleteMemoryAsync(string memoryId, string collectionNa
             var keys = Enumerable.Range(0, chunkSize).Select(i => BuildMemoryIdentifier(memoryId, i));
             await MemoryStore.RemoveBatchAsync(collectionName, keys, cancellationToken);
-            RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.RemoveBatch, Keys = keys, CollectionName = collectionName });
+            RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.RemoveBatch, Keys = [memoryId], CollectionName = collectionName });
```
I think this is not correct. If you run a batch the keys should be returned. If you want to return the memoryId, maybe it would be better to add one more property to the event model.

rliberoff replied:
Changed as agreed.

fjbelizon commented on src/Encamina.Enmarcha.SemanticKernel/MemoryStoreExtender.cs:
```diff
@@ -77,7 +77,7 @@ public virtual async Task<MemoryContent> GetMemoryAsync(string memoryId, string
         var keys = Enumerable.Range(0, chunkSize).Select(i => BuildMemoryIdentifier(memoryId, i));
         var memoryRecords = await MemoryStore.GetBatchAsync(collectionName, keys, cancellationToken: cancellationToken).ToListAsync(cancellationToken);
-        RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.GetBatch, Keys = keys, CollectionName = collectionName });
+        RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.GetMemory, Keys = [memoryId], CollectionName = collectionName });
```
Same as previous

rliberoff replied:
Changed as agreed.

fjbelizon commented on src/Encamina.Enmarcha.SemanticKernel/MemoryStoreExtender.cs:
```diff
@@ -91,6 +91,8 @@ public async Task<bool> ExistsMemoryAsync(string memoryId, string collectionName
     {
         var chunkSize = await GetChunkSize(memoryId, collectionName, cancellationToken);
 
+        RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.ExistsMemory, Keys = [memoryId], CollectionName = collectionName });
```
Same as previous

rliberoff replied:
Changed as agreed.

fjbelizon commented on src/Encamina.Enmarcha.SemanticKernel/MemoryStoreExtender.cs:
```diff
@@ -158,8 +160,6 @@ private async Task<int> GetChunkSize(string memoryId, string collectionName, Can
             /* Do nothing */
         }
 
-        RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.Get, Keys = [key], CollectionName = collectionName });
```
Same as previous

rliberoff replied:
Changed as agreed.

LuisM000 commented on src/Encamina.Enmarcha.AspNet.OpenApi/Constants.cs:
```diff
@@ -0,0 +1,18 @@
+        /// <summary>
+        /// The OpenAPI key header.
+        /// </summary>
+        public static readonly string KeyHeader = @"x-openapi-key";
```
Could it be a constant?

rliberoff replied:
That is a good question.
When you have a public constants class with members that could be used from the outside, it is more convenient to have those constants as `static readonly` rather than `const`. The explanation is a bit tricky, so I leave here some sites that explain it better than me 😅

  - https://lassiautio.com/2018/01/21/constant-or-static-readonly/
  - https://next.sonarqube.com/sonarqube/coding_rules?rule_key=csharpsquid%3AS2339&open=csharpsquid%3AS2339

LuisM000 commented on src/Encamina.Enmarcha.AspNet.OpenApi/README.md:
```diff
@@ -0,0 +1,66 @@
+
+#### GroupNameKeyAuthorizationMiddleware
+
+This middleware is used to block access to an OpenAPI document or specification b y using a pre-shared key that must be provided in the request header per specification.
```
Perhaps we intended to write 'by' without a space.

rliberoff replied:
Fixed.

rliberoff committed:
Fixed typo in `README.md` file of `Encamina.Enmarcha.AspNet.OpenApi`.

rliberoff committed:
Fixed referente to `aspnetcore-8.0` in `README.md` of `Encamina.Enmarcha.AspNet.OpenApi.Swashbuckle`.

rliberoff committed:
Removed default `null` for `setupAction` parameter in `UseOpenApiGroupNameKeyAuthorization` extension method.

rliberoff committed:
From PR comments, new property `MemoryIds` to replace `Keys` in `MemoryStoreEventArgs`. Now, `Keys` is deprecated.

rliberoff committed:
- Updated dependencies:
  - Updated `Azure.AI.OpenAI` from `1.0.0-beta12` to `1.0.0-beta13` (which provides some fixes for Function Calling).
  - Updated `Azure.Data.Tables` from `12.8.2` to `12.8.3`.
  - Updated `Microsoft.Bot.Builder.Azure` from `4.22.0` to `4.22.1`.
  - Updated `Microsoft.Bot.Builder.Azure.Blobs` from `4.22.0` to `4.22.1`.
  - Updated `Microsoft.Bot.Builder.Dialogs` from `4.22.0` to `4.22.1`.
  - Updated `Microsoft.Bot.Builder.Integration.ApplicationInsights.Core` from `4.22.0` to `4.22.1`.
  - Updated `MMicrosoft.Bot.Builder.Integration.AspNet.Core` from `4.22.0` to `4.22.1`.
  - Updated `Microsoft.Semantic Kernel` from `1.3.0` to `1.3.1`.
  - Updated `Microsoft.SemanticKernel.Connectors.AzureAISearch` from `1.3.0-alpha` to `1.3.1-alpha`. **Important**: this does not fix the issue detected by ENMARCHA and reported in [Issue 72](https://github.com/Encamina/enmarcha/issues/72).
  - Updated `Microsoft.SemanticKernel.Connectors.Qdrant` from `1.3.0-alpha` to `1.3.1-alpha`.
  - Updated `Microsoft.SemanticKernel.Plugins.Document` from `1.3.0-alpha` to `1.3.1-alpha`.
  - Updated `Microsoft.SemanticKernel.Plugins.Memory` from `1.3.0-alpha` to `1.3.1-alpha`.
  - Updated `SharpToken` from `1.2.14` to `1.2.15`.
